[IsDeprecatedWeakRefSmartPointerException] Make WebProcess/XR and WebProcess/WebStorage files ref counted

RupinMittal opened the pull request:
#### 608e68962f08d8e1b0e05c26788b0b08e9552981
<pre>
[IsDeprecatedWeakRefSmartPointerException] Make WebProcess/XR and WebProcess/WebStorage files ref counted
<a href="https://bugs.webkit.org/show_bug.cgi?id=280231">https://bugs.webkit.org/show_bug.cgi?id=280231</a>
<a href="https://rdar.apple.com/136536190">rdar://136536190</a>

Reviewed by Chris Dumez.

Remove IsDeprecatedWeakRefSmartPointerException. Since PlatformXRSystemProxy
has a single owner which does not re-assign it, we make PlatformXRSystemProxy
ref counted by adding ref() and deref() functions that forward the ref&apos;ing to
the owner rather than making PlatformXRSystemProxy a subclass of RefCounted&lt;&gt;.

That&apos;s not the case for StorageAreaMap, so we make it subclass RefCounted.

* Source/WebKit/Shared/XR/XRDeviceProxy.cpp:
(WebKit::XRDeviceProxy::initializeTrackingAndRendering):
(WebKit::XRDeviceProxy::shutDownTrackingAndRendering):
(WebKit::XRDeviceProxy::didCompleteShutdownTriggeredBySystem):
(WebKit::XRDeviceProxy::requestFrame):
(WebKit::XRDeviceProxy::createLayerProjection):
(WebKit::XRDeviceProxy::submitFrame):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::xrSystemProxy):
* Source/WebKit/WebProcess/WebProcess.cpp:
(WebKit::WebProcess::networkProcessConnectionClosed):
* Source/WebKit/WebProcess/WebStorage/StorageAreaImpl.cpp:
(WebKit::StorageAreaImpl::~StorageAreaImpl):
(WebKit::StorageAreaImpl::length):
(WebKit::StorageAreaImpl::key):
(WebKit::StorageAreaImpl::item):
(WebKit::StorageAreaImpl::setItem):
(WebKit::StorageAreaImpl::removeItem):
(WebKit::StorageAreaImpl::clear):
(WebKit::StorageAreaImpl::contains):
(WebKit::StorageAreaImpl::storageType const):
(WebKit::StorageAreaImpl::prewarm):
* Source/WebKit/WebProcess/WebStorage/StorageAreaMap.h:
* Source/WebKit/WebProcess/WebStorage/StorageNamespaceImpl.cpp:
(WebKit::StorageNamespaceImpl::storageArea):
* Source/WebKit/WebProcess/WebStorage/StorageNamespaceImpl.h:
* Source/WebKit/WebProcess/XR/PlatformXRSystemProxy.cpp:
(WebKit::PlatformXRSystemProxy::ref const):
(WebKit::PlatformXRSystemProxy::deref const):
* Source/WebKit/WebProcess/XR/PlatformXRSystemProxy.h:

Canonical link: <a href="https://commits.webkit.org/284189@main">https://commits.webkit.org/284189@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/78a8f9de7e049b9686a81ac8aaa7fb0601c4d1db

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/68652 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/48044 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/21311 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/72721 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/19797 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/55840 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/19613 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/54739 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/13168 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/71719 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/43905 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/59278 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/35204 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/40572 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/18154 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/62524 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/17049 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/74415 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/12623 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/16301 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/62216 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/12663 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/59357 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/62242 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/15249 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/10198 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/3813 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/43845 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/44919 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/46113 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/44661 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebKit/WebProcess/WebStorage/StorageNamespaceImpl.cpp:
```diff
@@ -81,7 +81,7 @@ void StorageNamespaceImpl::destroyStorageAreaMap(StorageAreaMap& map)
 {
     auto& map = m_storageAreaMaps.ensure(securityOrigin.data(), [&] {
-        return makeUnique<StorageAreaMap>(*this, securityOrigin);
+        return makeUniqueWithoutRefCountedCheck<StorageAreaMap>(*this, securityOrigin);
```
Forwarding ref-counting here doesn't seem safe. There is code removing the StorageAreaMap from m_storageAreaMaps, which would destroy the StorageAreaMap even if someone was holding a RefPtr to it.

RupinMittal replied:
Good point--it should be fixed now.

cdumez commented on Source/WebKit/WebProcess/WebProcess.cpp:
```diff
@@ -1270,7 +1270,7 @@ void WebProcess::networkProcessConnectionClosed(NetworkProcessConnection* connec
 
     for (auto key : copyToVector(m_storageAreaMaps.keys())) {
-        if (auto map = m_storageAreaMaps.get(key))
+        if (RefPtr map = m_storageAreaMaps.get(key).get())
```
do we really need the `.get()` here?

RupinMittal replied:
I believe we do. Is there some other safe way to convert a WeakPtr to a RefPtr?

cdumez replied:
Did you try?

cdumez replied:
I'm not sure why we're talking about a WeakPtr here.

The map is:
```
HashMap<WebCore::SecurityOriginData, Ref<StorageAreaMap>> m_storageAreaMaps;
```

The value is a `Ref`, when you call `get()` on the HashMap, you'd get a `RefPtr`. So the type already matches, and I suspect we don't need `.get()`

whsieh replied:
The map is on `WebProcess`:

```
HashMap<StorageAreaMapIdentifier, WeakPtr<StorageAreaMap>> m_storageAreaMaps;
```

(Are you suggesting the `WeakPtr` should be changed to a `RefPtr`?)

cdumez replied:
No, I'm certainly not suggesting switching from WeakPtr to RefPtr. I must have looked up the wrong HashMap somehow due to similar naming. The one I was looking at had a RefPtr as value. Ignore my comment then.

cdumez reviewed (approved):
LGTM

RupinMittal committed:
[IsDeprecatedWeakRefSmartPointerException] Make WebProcess/XR and WebProcess/WebStorage files ref counted
https://bugs.webkit.org/show_bug.cgi?id=280231
rdar://136536190

Reviewed by Chris Dumez.

Remove IsDeprecatedWeakRefSmartPointerException. Since PlatformXRSystemProxy
has a single owner which does not re-assign it, we make PlatformXRSystemProxy
ref counted by adding ref() and deref() functions that forward the ref'ing to
the owner rather than making PlatformXRSystemProxy a subclass of RefCounted<>.

That's not the case for StorageAreaMap, so we make it subclass RefCounted.

* Source/WebKit/Shared/XR/XRDeviceProxy.cpp:
(WebKit::XRDeviceProxy::initializeTrackingAndRendering):
(WebKit::XRDeviceProxy::shutDownTrackingAndRendering):
(WebKit::XRDeviceProxy::didCompleteShutdownTriggeredBySystem):
(WebKit::XRDeviceProxy::requestFrame):
(WebKit::XRDeviceProxy::createLayerProjection):
(WebKit::XRDeviceProxy::submitFrame):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::xrSystemProxy):
* Source/WebKit/WebProcess/WebProcess.cpp:
(WebKit::WebProcess::networkProcessConnectionClosed):
* Source/WebKit/WebProcess/WebStorage/StorageAreaImpl.cpp:
(WebKit::StorageAreaImpl::~StorageAreaImpl):
(WebKit::StorageAreaImpl::length):
(WebKit::StorageAreaImpl::key):
(WebKit::StorageAreaImpl::item):
(WebKit::StorageAreaImpl::setItem):
(WebKit::StorageAreaImpl::removeItem):
(WebKit::StorageAreaImpl::clear):
(WebKit::StorageAreaImpl::contains):
(WebKit::StorageAreaImpl::storageType const):
(WebKit::StorageAreaImpl::prewarm):
* Source/WebKit/WebProcess/WebStorage/StorageAreaMap.h:
* Source/WebKit/WebProcess/WebStorage/StorageNamespaceImpl.cpp:
(WebKit::StorageNamespaceImpl::storageArea):
* Source/WebKit/WebProcess/WebStorage/StorageNamespaceImpl.h:
* Source/WebKit/WebProcess/XR/PlatformXRSystemProxy.cpp:
(WebKit::PlatformXRSystemProxy::ref const):
(WebKit::PlatformXRSystemProxy::deref const):
* Source/WebKit/WebProcess/XR/PlatformXRSystemProxy.h:

Canonical link: https://commits.webkit.org/284189@main
